contracts/verifiers/Verifier.sol

```diff
@@ -745,16 +713,85 @@ abstract contract Verifier is IVerifier, ContextUpgradeable {
         return (false, 0);
     }
 
-    function _checkGroupsRequestsCount(
+    function _checkGroupsRequestsInfo(
         uint256[] memory groupList,
-        uint256[] memory groupRequestsList,
+        uint256[] memory groupRequestCountList,
+        bool[] memory groupUserIDInRequestsList,
         uint256 groupsCount
     ) internal pure {
         for (uint256 i = 0; i < groupsCount; i++) {
-            if (groupRequestsList[i] < 2) {
+            if (groupRequestCountList[i] < 2) {
                 revert GroupMustHaveAtLeastTwoRequests(groupList[i]);
             }
+            if (groupUserIDInRequestsList[i] == false) {
+                revert MissingUserIDInGroupOfRequests(groupList[i]);
+            }
+        }
+    }
+
+    function _checkGroupIdsAndRequestsPerGroup(
+        IVerifier.Request[] calldata requests
+    ) internal view {
+        uint256 newGroupsCount = 0;
+        uint256[] memory newGroupsGroupID = new uint256[](requests.length);
+        uint256[] memory newGroupsRequestCount = new uint256[](requests.length);
+        bool[] memory newGroupsUserIDInRequests = new bool[](requests.length);
+
+        for (uint256 i = 0; i < requests.length; i++) {
+            uint256 groupID = requests[i]
+            .validator
+            .getRequestParams(requests[i].params)[
+                requests[i].validator.requestParamIndexOf("groupID")
+            ].value;
+
+            if (groupID != 0) {
+                if (groupIdExists(groupID)) {
+                    revert GroupIdAlreadyExists(groupID);
+                }
+
+                (bool exists, uint256 groupIDIndex) = _getGroupIDIndex(
+                    groupID,
+                    newGroupsGroupID,
+                    newGroupsCount
+                );
+
+                if (!exists) {
+                    newGroupsGroupID[newGroupsCount] = groupID;
+                    newGroupsRequestCount[newGroupsCount]++;
+                    newGroupsCount++;
+                } else {
+                    newGroupsRequestCount[groupIDIndex]++;
+                }
+
+                if (_isUserIDInputInRequest(requests[i])) {
+                    newGroupsUserIDInRequests[groupIDIndex] = true;
+                }
+            } else {
+                // request without group
+                if (!_isUserIDInputInRequest(requests[i])) {
+                    revert MissingUserIDInRequest(requests[i].requestId);
+                }
+            }
         }
+
+        _checkGroupsRequestsInfo(
+            newGroupsGroupID,
+            newGroupsRequestCount,
+            newGroupsUserIDInRequests,
+            newGroupsCount
+        );
+    }
+
+    function _isUserIDInputInRequest(
+        IVerifier.Request memory request
+    ) internal view returns (bool) {
+        bool userIDInRequests = false;
+
+        try request.validator.inputIndexOf("userID") {
```

Comment:
Can you please bring constant instead of "userID"?

Reply:
Added also precalculated keccack256 for name `userID`